ujt/state.py

```diff
@@ -349,3 +359,111 @@ def set_node_override_status(
     else:
         override_status_map[node_name] = override_status
     cache.set("override_status_map", override_status_map)
+
+
+#@cache.memoize()
+def get_node_to_user_journey_map() -> Dict[str, List[UserJourney]]:
+    # map the node name to user journey names that pass through the node
+    # should this be in state? it's memoized but doesn't really affect state
+    # however, it's kind of similar to the other maps we expose in state.py,
+    # only this one is dynamically generated once...
+    node_name_message_map, client_name_message_map = get_message_maps()
+    output_map: DefaultDict[
+        str,
+        List[UserJourney]] = defaultdict(
+            list
+        )  # we would prefer to use a set here, but protobufs are not hashable
```

Comment:
UserJourney does have a unique identifier, which is ClientName+Name. So one option would be to make a map of UserJourneyId->UserJourney, where UserJourneyId is that combined string. Then build the map of Node to Set\<UserJouneyId\>. This keeps you from having to do a bunch of existence checking. Then as a final step, merge those two into a map of Node -> List\<UserJourney\>.

Step one is O(UserJourneys), step 2 is basically what you're doing already but without the dedup looping, and step 3 is O(total entries in the output of step 2)

Whether or not that's more efficient really depends on your ratio of Journeys to Nodes; I would say it's more elegant.

Reply:
Thanks for the feedback -- I hadn't considered that method initially. I agree that it's more elegant (and more obvious what each step is doing). I think it'll actually be more efficient too since our dedup checks take O(nodes) right now (and it would be reduced to O(1)).

Ideally I'd like to add a commit to this branch to fix this there's no way to automatically rebase all branches after this onto the new commit. Created an issue for it (#53) so I can fix it in a later branch.